### ISSUE TITLE
Rename extract method because it conflicts with `tidyr::extract`

### ISSUE THREAD

**Issue by jrnold** [stan-dev/rstan#118]
The `extract` method shares a name with the `extract` function in `tidyr`. This isn't an issue at the moment, since `tidyr` is a fairly new package. But I wanted to put it on the radar when considering the next version of rstan.  Given that `tidyr` is one of Hadley Wickham's packages and is being pushed as a complement to `dplyr` for data cleaning and manipulation, it'll probably be one of the more widely used R packages within a year, and having a function that conflicts with one in that package will cause annoyance.

**Comment by casallas** [comment]
I wanted to point out that there is another name clash with `magrittr::extract`. `magrittr` which provides the pipe operator (`%>%`) in R is also becoming very popular (and is also authored by Hadley Wickham) and I've come across this problem a couple of times, my current solution is to load all the packages in the correct order (i.e. `library(magrittr);library(rstan)` at the top of my scripts to avoid weird errors.

Interestingly, magrittr's `x %>% extract(1)` is equivalent to `x[1]`, and `x %>% extract2(1)` is equivalent to `x[[1]]` which is close to what you guys were discussing!

**Comment by bob-carpenter** [comment]
[cc-ing stan-users]

Thanks.  We've known about some pending conflicts for
a while now.

The right thing to do from a programming point of view
is use namespaces.  R's defaults are very badly behaved 
from a conventional programming point of view, because 
they load all the namespaces into the top level.

I don't believe namespaces are widely known among R 
programmers.  Hadley has a tutorial:

http://r-pkgs.had.co.nz/namespace.html

In short, you should be able to use tidyr::extract for the 
tidy version and rstan::extract for the Stan version.  
- Bob

**Comment by lionel-** [comment]
> I find the behavior or returning a list when only a single variable is requested to be inconsistent with R's admittedly weird dropping of dimensions.

This convention must go away! Don't feed it :)

dplyr's `[.tbl_df` method does not drop dimensions for example.

**Comment by lionel-** [comment]
A serious issue with rstan related to this is that it's not possible to call `stan()` as follows:

``` r
rstan::stan(...)
```

because it automatically attaches two packages: inline and rstan. This is really problematic because it imposes all rstan's exported functions in the search path. This can provoke name clashes with previously loaded packages. It's really not good practice to impose attachment of packages.

So, is it really necessary to attach these two packages everytime we fire up `stan()`?

**Comment by bgoodri** [comment]
The `extract` symbol name won't go away before rstan 3.0 but it probably will go away for 3.0.
There will probably be [ and [[ methods for 3.0
I think one of the reasons why rstan::stan() doesn't work is that Rcpp is necessary but not loaded.

**Comment by jgabry** [comment]
I'm closing this issue because we're not going to worry about the "extract" conflict. When we introduce the new R interface 

https://github.com/stan-dev/stan/wiki/User-Interface-Guidelines-for-Developers 

it won't be an issue anymore.
